keymanager/src/crypto/kdf.rs

```diff
@@ -419,30 +429,21 @@ impl Kdf {
     }
 
     /// Get the public part of the key.
-    pub fn get_public_key(&self, req: &impl KeyRequest) -> Result<Option<PublicKey>> {
+    pub fn get_public_key(&self, req: &impl KeyRequest) -> Result<PublicKey> {
         let keys = self.get_or_create_keys(req)?;
-        Ok(Some(keys.input_keypair.pk))
+        Ok(keys.input_keypair.pk)
     }
 
     /// Signs the public key using the key manager key.
-    pub fn sign_public_key(&self, key: PublicKey) -> Result<SignedPublicKey> {
-        let mut body = key.as_ref().to_vec();
-
+    pub fn sign_public_key(&self, key: PublicKey, req: Vec<u8>) -> Result<SignedPublicKey> {
         let inner = self.inner.read().unwrap();
         let checksum = inner.get_checksum()?;
-        body.extend_from_slice(&checksum);
-
         let signer = inner
             .signer
             .as_ref()
             .ok_or(KeyManagerError::NotInitialized)?;
-        let signature = signer.sign(&PUBLIC_KEY_CONTEXT, &body)?;
 
-        Ok(SignedPublicKey {
-            key,
-            checksum,
-            signature,
-        })
+        SignedPublicKey::new(key, checksum, req, signer)
```

Comment:
So `req: Vec<u8>` is a sort of cryptographic context to indicate what the signature "means", but it's not perfect. In particular, `req` is a serialized request, presumably by oasis-cbor, and the serialized struct obeys canonical cbor like rules with struct member names used as keys in a map, and keys are sorted.  this means that two structs with different names/purposes but with identical members (names and types, but possibly occuring in a different order) will serialize into the same "type" of byte layout, even though rust will treat the two structs as different types -- the encoded bytes of one struct's value could be decoded as an instance of the other struct without oasis-cbor generating an error.  so if ever there were two requests that use same-cbor-layout structs, then the replies could be confused.  it's unlikely to occur, but it's a potential gotcha.